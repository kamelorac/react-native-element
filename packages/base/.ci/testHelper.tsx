import React, { JSXElementConstructor } from 'react';
import { View } from 'react-native';
import {
  render,
  RenderOptions,
  fireEvent,
  act,
} from '@testing-library/react-native';

export { fireEvent, act };

// for getting findByType e.g. wrapper.findByType(Icon) see implementation in Avatar Component
export const renderWithWrapper = (
  children: React.ReactElement<any, string | JSXElementConstructor<any>>,
  wrapperTestID?: string,
  _themeProp: unknown = {},
  renderOptions?: RenderOptions
) => {
  const options: RenderOptions = {
    ...(!wrapperTestID && {
      wrapper: (props) => <View {...props} testID="wrapper" />,
    }),
    ...renderOptions,
  };
  const renderApi = render(children, options);
  const wrapper = renderApi.queryByTestId(wrapperTestID || 'wrapper')!;
  return { wrapper, ...renderApi };
};
