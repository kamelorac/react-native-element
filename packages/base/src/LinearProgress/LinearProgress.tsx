import React from 'react';
import {
  View,
  StyleProp,
  ViewStyle,
  Animated,
  Platform,
  ViewProps,
} from 'react-native';
import Color from 'color';
import { defaultTheme, RneFunctionComponent } from '../helpers';

export interface LinearProgressProps extends ViewProps {
  /** The value of the progress indicator for the determinate variant. Value between 0 and 1. */
  value?: number;

  /** Type of button. */
  variant?: 'determinate' | 'indeterminate';

  /** Color for linear progress. */
  color?: 'primary' | 'secondary' | string;

  /** Track color for linear progress. */
  trackColor?: string;

  /** Add additional styling for linear progress component. */
  style?: StyleProp<ViewStyle>;

  /** Animation duration */
  animation?:
    | {
        duration?: number;
      }
    | boolean;
}

/** Progress indicators inform users about the status of ongoing processes, such as loading an app, submitting a form, or saving updates.
 * They communicate an appâ€™s state and indicate available actions, such as whether users can navigate away from the current screen.
 * Also receives all [View](https://reactnative.dev/docs/view#props) props */
export const LinearProgress: RneFunctionComponent<LinearProgressProps> = ({
  value,
  /** If value is given then variant default as determinate */
  variant = value === undefined ? 'indeterminate' : 'determinate',
  color = 'secondary',
  style,
  theme = defaultTheme,
  trackColor,
  animation = { duration: 2000 },
  ...rest
}) => {
  const [width, setWidth] = React.useState<number>(0);

  const { current: transition } = React.useRef<Animated.Value>(
    new Animated.Value(0)
  );

  const intermediate = React.useRef<Animated.CompositeAnimation>();

  const startAnimation = React.useCallback(() => {
    if (variant === 'indeterminate') {
      intermediate.current = Animated.timing(transition, {
        duration: typeof animation !== 'boolean' ? animation.duration : 2000,
        toValue: 1,
        useNativeDriver: true,
        isInteraction: false,
      });
      transition.setValue(0);
      Animated.loop(intermediate.current).start();
    } else {
      Animated.timing(transition, {
        duration: typeof animation !== 'boolean' ? animation.duration : 1000,
        toValue: value || 0,
        useNativeDriver: Platform.OS !== 'web',
        isInteraction: false,
      }).start();
    }
  }, [variant, transition, animation, value]);

  const tintColor =
    color === 'secondary' || color === 'primary'
      ? theme?.colors?.[color]
      : Color(color).rgb().string() || theme?.colors?.secondary;

  const trackTintColor =
    trackColor || Color(tintColor).alpha(0.4).rgb().string();

  React.useEffect(() => {
    if (animation) {
      startAnimation();
    }
  }, [animation, startAnimation]);

  return (
    <View
      testID="RNE__LinearProgress_Bar"
      accessible
      accessibilityRole="progressbar"
      accessibilityValue={{
        now: clamp(value),
        min: 0,
        max: 1,
      }}
      {...rest}
      onLayout={(e) => {
        setWidth(e.nativeEvent.layout.width);
      }}
      style={[
        {
          height: 4,
          overflow: 'hidden',
          width: '100%',
          backgroundColor: trackTintColor,
          position: 'relative',
        },
        style,
      ]}
    >
      {animation ? (
        <Animated.View
          testID="RNE__LinearProgress_Progress"
          style={{
            transform: [
              {
                translateX: transition.interpolate(
                  variant === 'indeterminate'
                    ? {
                        inputRange: [0, 1],
                        outputRange: [-width, 0.5 * width],
                      }
                    : {
                        inputRange: [0, 1],
                        outputRange: [-0.5 * width, 0],
                      }
                ),
              },
              {
                scaleX: transition.interpolate(
                  variant === 'indeterminate'
                    ? {
                        inputRange: [0, 0.5, 1],
                        outputRange: [0.0001, 1, 0.001],
                      }
                    : {
                        inputRange: [0, 1],
                        outputRange: [0.0001, 1],
                      }
                ),
              },
            ],
            backgroundColor: tintColor as string,
            flex: 1,
          }}
        />
      ) : (
        <View
          testID="RNE__LinearProgress_Progress"
          style={{
            flex: 1,
            width: width * clamp(value || 0),
            backgroundColor: tintColor as string,
          }}
        />
      )}
    </View>
  );
};

LinearProgress.displayName = 'LinearProgress';

/**
 * Keep value between 0 and 1
 */
export const clamp = (value: number): number =>
  Math.max(0, Math.min(value, 1)) || 0;
