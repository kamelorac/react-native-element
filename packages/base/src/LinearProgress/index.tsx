import {
  LinearProgress as LinearProgressBase,
  LinearProgressProps,
} from './LinearProgress';

/**
 * ### Linear Progress
 *
 * Progress indicators inform users about the status of ongoing processes, such as loading an app, submitting a form, or saving updates. They communicate an appâ€™s state and indicate available actions, such as whether users can navigate away from the current screen.
 *
 * Usage ```<LinearProgress variant={LinearProgress.INDETERMINATE}  />```
 *
 * [API Documentation](https://reactnativeelements.com/docs/linearProgress)
 */
const LinearProgress = Object.assign(LinearProgressBase, {
  INDETERMINATE: 'indeterminate' as LinearProgressProps['variant'],
  DETERMINATE: 'determinate' as LinearProgressProps['variant'],
});

export { LinearProgress };
export type { LinearProgressProps };
