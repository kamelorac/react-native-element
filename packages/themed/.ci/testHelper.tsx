import React, { JSXElementConstructor } from 'react';
import { View } from 'react-native';
import {
  render,
  RenderOptions,
  fireEvent,
  act,
} from '@testing-library/react-native';
import {
  ThemeProvider,
  FullTheme,
  createTheme,
  CreateThemeOptions,
} from '../src/config';

export { fireEvent, act };

// for getting findByType e.g. wrapper.findByType(Icon) see implementation in Avatar Component
export const renderWithWrapper = (
  children: React.ReactElement<any, string | JSXElementConstructor<any>>,
  wrapperTestID?: string,
  themeProp: Partial<CreateThemeOptions> = {},
  renderOptions?: RenderOptions
) => {
  const options: RenderOptions = {
    ...(!wrapperTestID && {
      wrapper: (props) => <View {...props} testID="wrapper" />,
    }),
    ...renderOptions,
  };
  const renderApi = render(
    <ThemeProvider theme={createTheme(themeProp)}>{children}</ThemeProvider>,
    options
  );
  const wrapper = renderApi.queryByTestId(wrapperTestID || 'wrapper')!;
  return { wrapper, ...renderApi };
};
