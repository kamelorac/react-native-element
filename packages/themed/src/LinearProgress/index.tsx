import { withTheme } from '../config';
import {
  LinearProgress,
  LinearProgressProps,
} from '@rneui/base/dist/LinearProgress/LinearProgress';

export { LinearProgress };
export type { LinearProgressProps };
/**
 * ### Linear Progress
 *
 * Progress indicators inform users about the status of ongoing processes, such as loading an app, submitting a form, or saving updates. They communicate an appâ€™s state and indicate available actions, such as whether users can navigate away from the current screen.
 *
 * Usage ```<LinearProgress variant={LinearProgress.INDETERMINATE}  />```
 *
 * [API Documentation](https://reactnativeelements.com/docs/linearProgress)
 */
export default Object.assign(
  withTheme<LinearProgressProps>(LinearProgress, 'LinearProgress'),
  {
    INDETERMINATE: 'indeterminate' as LinearProgressProps['variant'],
    DETERMINATE: 'determinate' as LinearProgressProps['variant'],
  }
);
