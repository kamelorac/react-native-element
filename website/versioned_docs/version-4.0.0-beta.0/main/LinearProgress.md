---
id: linearprogress

title: LinearProgress

slug: /linearprogress
---

import Usage from './usage/LinearProgress/LinearProgress.md'

Progress indicators inform users about the status of ongoing processes, such as loading an app, submitting a form, or saving updates.

They communicate an appâ€™s state and indicate available actions, such as whether users can navigate away from the current screen.

Also receives all [View](https://reactnative.dev/docs/view#props) props

## Usage

<Usage />

---

## Props

### LinearProgress

- [animation](#animation)
- [color](#color)
- [style](#style)
- [trackColor](#trackcolor)
- [value](#value)
- [variant](#variant)

## Reference

### LinearProgress

#### animation

Animation duration

| Type                              | Default            |
| --------------------------------- | ------------------ |
| boolean or { duration?: number; } | { duration: 2000 } |

---

#### color

Color for linear progress.

| Type   | Default   |
| ------ | --------- |
| string | secondary |

---

#### style

Add additional styling for linear progress component.

| Type               | Default |
| ------------------ | ------- |
| View style(Object) | None    |

---

#### trackColor

Track color for linear progress.

| Type   | Default |
| ------ | ------- |
| string | None    |

---

#### value

The value of the progress indicator for the determinate variant. Value between 0 and 1.

| Type   | Default |
| ------ | ------- |
| number | None    |

---

#### variant

Type of button.

| Type                             | Default                                               |
| -------------------------------- | ----------------------------------------------------- |
| "determinate" or "indeterminate" | value === undefined ? 'indeterminate' : 'determinate' |

---
