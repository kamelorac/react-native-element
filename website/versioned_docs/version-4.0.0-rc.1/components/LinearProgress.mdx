---
id: linearprogress
title: LinearProgress
---

import Tabs from "@theme/Tabs";
import TabItem from "@theme/TabItem";
import {} from "react-native-elements";
import Usage from "../component_usage/LinearProgress.mdx";

Progress indicators inform users about the status of ongoing processes, such as loading an app, submitting a form, or saving updates.
They communicate an appâ€™s state and indicate available actions, such as whether users can navigate away from the current screen.
Also receives all [View](https://reactnative.dev/docs/view#props) props

## Usage

<Usage />

## Props

:::note
Includes all [View](https://reactnative.dev/docs/view#props) props.
:::

<div class='table-responsive'>

| Name         | Type                                  | Default                                                 | Description                                                                             |
| ------------ | ------------------------------------- | ------------------------------------------------------- | --------------------------------------------------------------------------------------- |
| `animation`  | `boolean` or `{ duration?: number; }` | `{ duration: 2000 }`                                    | Animation duration                                                                      |
| `color`      | string                                | `secondary`                                             | Color for linear progress.                                                              |
| `style`      | View Style                            |                                                         | Add additional styling for linear progress component.                                   |
| `trackColor` | string                                |                                                         | Track color for linear progress.                                                        |
| `value`      | number                                |                                                         | The value of the progress indicator for the determinate variant. Value between 0 and 1. |
| `variant`    | `determinate` or `indeterminate`      | `value === undefined ? 'indeterminate' : 'determinate'` | Type of button.                                                                         |

</div>
