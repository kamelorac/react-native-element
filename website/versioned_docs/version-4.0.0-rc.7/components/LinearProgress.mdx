---
id: linearprogress
title: LinearProgress
---

import { IconsStyle } from "@site/src/components/Docs_Icons";
import Tabs from "@theme/Tabs";
import TabItem from "@theme/TabItem";
import CodeBlock from "@theme/CodeBlock";
import { BiInfoCircle } from "react-icons/bi";
import Usage from "../component_usage/LinearProgress.mdx";
import Play from "@site/playground/LinearProgress/linearprogress.playground";

<IconsStyle />

Progress indicators inform users about the status of ongoing processes, such as loading an app, submitting a form, or saving updates.
They communicate an appâ€™s state and indicate available actions, such as whether users can navigate away from the current screen.
Also receives all [View](https://reactnative.dev/docs/view#props) props

## Usage

<div class="row inline-flex-center">
  <div class="col col--3">
    <h4>Import</h4>
  </div>
  <div class="col col--9">
    <CodeBlock>{`import { LinearProgress } from '@rneui/themed';`}</CodeBlock>
  </div>
  <div class="col col--3">
    <h4>
      Theme Key{" "}
      <a href="../customizing#using-themeprovider">
        <BiInfoCircle />
      </a>
    </h4>
  </div>
  <div class="col col--9">
    <CodeBlock>{`LinearProgress`}</CodeBlock>
  </div>
</div>

<Usage />

## Props

:::note
Includes all [View](https://reactnative.dev/docs/view#props) props.
:::

<div class='table-responsive'>

| Name         | Type                                  | Default                                                 | Description                                                                             |
| ------------ | ------------------------------------- | ------------------------------------------------------- | --------------------------------------------------------------------------------------- |
| `animation`  | `boolean` \| `{ duration?: number; }` | `{ duration: 2000 }`                                    | Animation duration                                                                      |
| `color`      | string                                | `secondary`                                             | Color for linear progress.                                                              |
| `style`      | View Style                            |                                                         | Add additional styling for linear progress component.                                   |
| `trackColor` | string                                |                                                         | Track color for linear progress.                                                        |
| `value`      | number                                |                                                         | The value of the progress indicator for the determinate variant. Value between 0 and 1. |
| `variant`    | `determinate` \| `indeterminate`      | `value === undefined ? 'indeterminate' : 'determinate'` | Type of button.                                                                         |

</div>

## Playground

<Play />
