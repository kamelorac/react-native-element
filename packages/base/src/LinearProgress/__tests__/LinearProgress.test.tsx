import React from 'react';
import { LinearProgress } from '../index';
import { renderWithWrapper, fireEvent, act } from '../../../.ci/testHelper';
import { clamp } from '../LinearProgress';

describe('LinearProgress Component', () => {
  it('should clamp', () => {
    expect(clamp(3)).toBe(1);
    expect(clamp(undefined)).toBe(0);
    expect(clamp(-1)).toBe(0);
    expect(clamp(0.6)).toBe(0.6);
  });
  it('should have apply color', () => {
    const { wrapper } = renderWithWrapper(
      <LinearProgress color="red" trackColor="blue" />,
      'RNE__LinearProgress_Bar'
    );
    expect(wrapper.props.style[0]).toMatchObject({
      height: 4,
      overflow: 'hidden',
      width: '100%',
      backgroundColor: 'blue',
      position: 'relative',
    });
  });

  it('should render determinant variant', () => {
    const component = renderWithWrapper(
      <LinearProgress variant={LinearProgress.DETERMINATE} value={0.4} />
    );
    expect(component.toJSON()).toMatchSnapshot();
  });

  it('should render inDeterminant variant', () => {
    const component = renderWithWrapper(
      <LinearProgress variant={LinearProgress.INDETERMINATE} />
    );
    expect(component.toJSON()).toMatchSnapshot();
  });

  it('should default to determinant if value is provided', () => {
    const { wrapper } = renderWithWrapper(
      <LinearProgress value={0.4} />,
      'RNE__LinearProgress_Progress'
    );
    act(() => {
      fireEvent(wrapper, 'layout', {
        nativeEvent: {
          layout: {
            width: 300,
          },
        },
      });
    });
    expect(wrapper.props.style.transform[0].translateX).toBe(300 * -0.5);
  });

  it('should render determinant variant without animation', () => {
    const { wrapper } = renderWithWrapper(
      <LinearProgress
        variant={LinearProgress.DETERMINATE}
        value={0.4}
        animation={false}
      />,
      'RNE__LinearProgress_Progress'
    );
    act(() => {
      fireEvent(wrapper, 'layout', {
        nativeEvent: {
          layout: {
            width: 300,
          },
        },
      });
    });
    expect(wrapper.props.style).toMatchObject({
      width: 300 * 0.4,
    });
  });

  it('should contain the required accessibility properties', () => {
    const component = renderWithWrapper(<LinearProgress value={0.4} />);
    const progressBar = component.getByRole('progressbar');
    expect(progressBar.props).toMatchObject({
      accessibilityValue: {
        now: 0.4,
        min: 0,
        max: 1,
      },
    });
  });
});
