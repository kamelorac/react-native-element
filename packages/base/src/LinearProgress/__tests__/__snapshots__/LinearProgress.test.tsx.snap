// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LinearProgress Component should render determinant variant 1`] = `
<View
  testID="wrapper"
>
  <View
    accessibilityRole="progressbar"
    accessibilityValue={
      {
        "max": 1,
        "min": 0,
        "now": 0.4,
      }
    }
    accessible={true}
    onLayout={[Function]}
    style={
      [
        {
          "backgroundColor": "rgba(173, 20, 87, 0.4)",
          "height": 4,
          "overflow": "hidden",
          "position": "relative",
          "width": "100%",
        },
        undefined,
      ]
    }
    testID="RNE__LinearProgress_Bar"
  >
    <View
      collapsable={false}
      style={
        {
          "backgroundColor": "#ad1457",
          "flex": 1,
          "transform": [
            {
              "translateX": -0,
            },
            {
              "scaleX": 0.0001,
            },
          ],
        }
      }
      testID="RNE__LinearProgress_Progress"
    />
  </View>
</View>
`;

exports[`LinearProgress Component should render inDeterminant variant 1`] = `
<View
  testID="wrapper"
>
  <View
    accessibilityRole="progressbar"
    accessibilityValue={
      {
        "max": 1,
        "min": 0,
        "now": 0,
      }
    }
    accessible={true}
    onLayout={[Function]}
    style={
      [
        {
          "backgroundColor": "rgba(173, 20, 87, 0.4)",
          "height": 4,
          "overflow": "hidden",
          "position": "relative",
          "width": "100%",
        },
        undefined,
      ]
    }
    testID="RNE__LinearProgress_Bar"
  >
    <View
      collapsable={false}
      style={
        {
          "backgroundColor": "#ad1457",
          "flex": 1,
          "transform": [
            {
              "translateX": -0,
            },
            {
              "scaleX": 0.0001,
            },
          ],
        }
      }
      testID="RNE__LinearProgress_Progress"
    />
  </View>
</View>
`;
